Add terraformtype Support for github.com/hashicorp/terraform-plugin-sdk/terraform.ResourceProvider

Version 2 of the Terraform Plugin SDK removes this interface type:

```
aws/provider.go:13:17: undefined: terraform.ResourceProvider
```

To support downstream Analyzers wishing to handle this, we can add constants and helper functions in a `helper/terraformtype/terraform` package similar to the others.

--- a/helper/terraformtype/terraform/package.go
+++ b/helper/terraformtype/terraform/package.go
@@ -1,0 +1,23 @@
+package terraform
+
+import (
+	"go/ast"
+	"go/types"
+
+	"github.com/bflad/tfproviderlint/helper/astutils"
+)
+
+const (
+	PackageName = `terraform`
+	PackagePath = `github.com/hashicorp/terraform-plugin-sdk/terraform`
+)
+
+// IsFunc returns if the function call is in the resource package
+func IsFunc(e ast.Expr, info *types.Info, funcName string) bool {
+	return astutils.IsPackageFunc(e, info, PackagePath, funcName)
+}
+
+// IsNamedType returns if the type name matches and is from the helper/resource package
+func IsNamedType(t *types.Named, typeName string) bool {
+	return astutils.IsPackageNamedType(t, PackagePath, typeName)
+}

--- a/helper/terraformtype/terraform/type_resourceprovider.go
+++ b/helper/terraformtype/terraform/type_resourceprovider.go
@@ -1,0 +1,21 @@
+package terraform
+
+import (
+	"go/types"
+)
+
+const (
+	TypeNameResourceProvider = `ResourceProvider`
+)
+
+// IsTypeResourceProvider returns if the type is ResourceProvider from the terraform package
+func IsTypeResourceProvider(t types.Type) bool {
+	switch t := t.(type) {
+	case *types.Named:
+		return IsNamedType(t, TypeNameResourceProvider)
+	case *types.Pointer:
+		return IsTypeResourceProvider(t.Elem())
+	default:
+		return false
+	}
+}

--- a/helper/terraformtype/terraform/type_resourceprovider_test.go
+++ b/helper/terraformtype/terraform/type_resourceprovider_test.go
@@ -1,0 +1,68 @@
+package terraform
+
+import (
+	"fmt"
+	"go/token"
+	"go/types"
+	"testing"
+)
+
+func TestIsTypeResourceProvider(t *testing.T) {
+	testCases := []struct {
+		Name      string
+		InputType types.Type
+		Expected  bool
+	}{
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, TypeNameResourceProvider),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameResourceProvider, nil),
+				nil,
+				nil,
+			),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("*%s.%s", PackagePath, TypeNameResourceProvider),
+			InputType: types.NewPointer(types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameResourceProvider, nil),
+				nil,
+				nil,
+			)),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, "Not"),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), "Not", nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", "incorrect/path", TypeNameResourceProvider),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage("incorrect/path", "path"), TypeNameResourceProvider, nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name:      "string",
+			InputType: types.Typ[types.String],
+			Expected:  false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			got := IsTypeResourceProvider(testCase.InputType)
+
+			if got != testCase.Expected {
+				t.Errorf("got %t, expected %t", got, testCase.Expected)
+			}
+		})
+	}
+}

